### ISSUE TITLE
Handle being closed too early? (ValueError: seek of closed file)

### ISSUE THREAD

**Issue by camFoltz** [cgohlke/tifffile#67]
I am try to load an .ome.tif stack that contains multiple series.  Using the master-ome-file I try to read the data with a few options:

```
tifffile.imread('/path_to_master')
tifffile.imread('/path_to_master', aszarr=True)
```
and 

```
positions = {}
with TiffFile(master_ome) as tif:
        for idx, tiffpageseries in enumerate(tif.series):
            positions[idx] = zarr.open(tiffpageseries.aszarr(), mode='r')
```

I am able to initialize a TiffFile object with `obj = TiffFile(master_ome)` but as soon as I try to access any information of the series such as `len(obj.series)` or I try to open with the above methods I get this error:

```
~/miniconda3/envs/waveOrder/lib/python3.7/site-packages/tifffile/tifffile.py in series(self)
   3176         ):
   3177             if getattr(self, 'is_' + name, False):
-> 3178                 series = getattr(self, '_series_' + name)()
   3179                 if not series and name == 'ome' and self.is_imagej:
   3180                     # try ImageJ series if OME series fails.

~/miniconda3/envs/waveOrder/lib/python3.7/site-packages/tifffile/tifffile.py in _series_ome(self)
   4079                     for i, keyframe in enumerate(ifds):
   4080                         if keyframe:
-> 4081                             keyframe.parent.pages.keyframe = keyframe.index
   4082                             keyframe = keyframe.parent.pages[keyframe.index]
   4083                             ifds[i] = keyframe

~/miniconda3/envs/waveOrder/lib/python3.7/site-packages/tifffile/tifffile.py in keyframe(self, index)
   4966         self._tiffpage = TiffPage
   4967         try:
-> 4968             self._keyframe = self._getitem(index)
   4969         finally:
   4970             self._tiffpage = tiffpage

~/miniconda3/envs/waveOrder/lib/python3.7/site-packages/tifffile/tifffile.py in _getitem(self, key, validate, cache, aspage)
   5229 
   5230         pageindex = key if self._index is None else self._index + (key,)
-> 5231         self._seek(key)
   5232         page = tiffpage(self.parent, index=pageindex, keyframe=self._keyframe)
   5233         if validate and validate != page.hash:

~/miniconda3/envs/waveOrder/lib/python3.7/site-packages/tifffile/tifffile.py in _seek(self, index, maxpages)
   5087         fh = self.parent.filehandle
   5088         if fh.closed:
-> 5089             raise ValueError('seek of closed file')
   5090 
   5091         if self._indexed or 0 <= index < lenpages:

ValueError: seek of closed file
```

I should note that I am trying to access the files locally, and that I am trying to do this through jupyter notebook, but this error also occurs if I run it through python in terminal.  I can also confirm that no other program is trying to access these files at the same time.

This was using the latest version of tifffile (2021.2.6)

Does anyone know what may be happening here?

**Comment by cgohlke** [comment]
I can't reproduce this with a single-file, multi-series OME TIFF. I don't have a multi-file, multi-series dataset. Could you please provide a standalone example and and share the files?

**Comment by camFoltz** [comment]
@cgohlke Here is a link to the dataset that was acquired by micromanager.  [https://drive.google.com/drive/folders/1_57ZflZIF9FS0IpuAK0TOlUtfv8SClgx?usp=sharing](https://drive.google.com/drive/folders/1_57ZflZIF9FS0IpuAK0TOlUtfv8SClgx?usp=sharing)

The dimensions of this data are: (Pos, Chan, Z, Y, X) = (4, 4, 1201, 512, 512)

We have been using this same build of MicroManager for a few months, but our data has only recently encountered this issue. This error only happens when we load data where the acquisition required more than one .ome.tif file to save it all.  Thanks for looking into this.

**Comment by cgohlke** [comment]
Got it. Thank you for the files. I can reproduce the issue and will look into it.

**Comment by cgohlke** [comment]
Should be fixed in v2021.3.4.

**Comment by jrussell25** [comment]
Hi there, Ive come across a similar error to this when trying to write a little test code that saves and loads tifffiles via zarr/dask. The following snippet generates the error.

```python
import numpy as np
import dask.array as da
import tifffile

shape = (5,4,6)
data = np.arange(np.prod(shape), dtype="uint16").reshape(shape)
tifffile.imsave("test.tif", data)

d_arr = da.from_zarr(tifffile.imread("test.tif", aszarr=True))
# breaks when computing the data  raises "ValueError: seek of closed file"
d_arr.compute()
```
<details>
  <summary>Full traceback</summary>
  
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-c853110d0a82> in <module>
      9 d_arr = da.from_zarr(tifffile.imread("test.tif", aszarr=True))
     10 # breaks when computing the data
---> 11 d_arr.compute()

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/base.py in compute(self, **kwargs)
    284         dask.base.compute
    285         """
--> 286         (result,) = compute(self, traverse=False, **kwargs)
    287         return result
    288 

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/base.py in compute(*args, **kwargs)
    566         postcomputes.append(x.__dask_postcompute__())
    567 
--> 568     results = schedule(dsk, keys, **kwargs)
    569     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    570 

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/threaded.py in get(dsk, result, cache, num_workers, pool, **kwargs)
     77             pool = MultiprocessingPoolExecutor(pool)
     78 
---> 79     results = get_async(
     80         pool.submit,
     81         pool._max_workers,

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/local.py in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    515                             _execute_task(task, data)  # Re-execute locally
    516                         else:
--> 517                             raise_exception(exc, tb)
    518                     res, worker_id = loads(res_info)
    519                     state["cache"][key] = res

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/local.py in reraise(exc, tb)
    323     if exc.__traceback__ is not tb:
    324         raise exc.with_traceback(tb)
--> 325     raise exc
    326 
    327 

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    221     try:
    222         task, data = loads(task_info)
--> 223         result = _execute_task(task, data)
    224         id = get_id()
    225         result = dumps((result, id))

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/core.py in _execute_task(arg, cache, dsk)
    119         # temporaries by their reference count and can execute certain
    120         # operations in-place.
--> 121         return func(*(_execute_task(a, cache) for a in args))
    122     elif not ishashable(arg):
    123         return arg

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/dask/array/core.py in getter(a, b, asarray, lock)
     97         lock.acquire()
     98     try:
---> 99         c = a[b]
    100         # Below we special-case `np.matrix` to force a conversion to
    101         # `np.ndarray` and preserve original Dask behavior for `getter`,

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/zarr/core.py in __getitem__(self, selection)
    660 
    661         fields, selection = pop_fields(selection)
--> 662         return self.get_basic_selection(selection, fields=fields)
    663 
    664     def get_basic_selection(self, selection=Ellipsis, out=None, fields=None):

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/zarr/core.py in get_basic_selection(self, selection, out, fields)
    785                                                 fields=fields)
    786         else:
--> 787             return self._get_basic_selection_nd(selection=selection, out=out,
    788                                                 fields=fields)
    789 

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/zarr/core.py in _get_basic_selection_nd(self, selection, out, fields)
    828         indexer = BasicIndexer(selection, self)
    829 
--> 830         return self._get_selection(indexer=indexer, out=out, fields=fields)
    831 
    832     def get_orthogonal_selection(self, selection, out=None, fields=None):

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/zarr/core.py in _get_selection(self, indexer, out, fields)
   1118 
   1119                 # load chunk selection into output array
-> 1120                 self._chunk_getitem(chunk_coords, chunk_selection, out, out_selection,
   1121                                     drop_axes=indexer.drop_axes, fields=fields)
   1122         else:

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/zarr/core.py in _chunk_getitem(self, chunk_coords, chunk_selection, out, out_selection, drop_axes, fields)
   1788         try:
   1789             # obtain compressed data for chunk
-> 1790             cdata = self.chunk_store[ckey]
   1791 
   1792         except KeyError:

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in __getitem__(self, key)
   8671         if key in self._store:
   8672             return self._store[key]
-> 8673         return self._getitem(key)
   8674 
   8675     def _getitem(self, key):

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in _getitem(self, key)
   9059     def _getitem(self, key):
   9060         """Return chunk from file."""
-> 9061         keyframe, page, chunkindex, offset, bytecount = self._parse_key(key)
   9062 
   9063         if self._chunkmode:

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in _parse_key(self, key)
   9124             return page.keyframe, page, None, None, None
   9125         else:
-> 9126             page = series[pageindex]
   9127             if page is None:
   9128                 return keyframe, None, chunkindex, 0, 0

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in __getitem__(self, key)
   8561         getitem = self._getitem
   8562         if isinstance(key, (int, numpy.integer)):
-> 8563             return getitem(key)
   8564         if isinstance(key, slice):
   8565             return [getitem(i) for i in range(*key.indices(self._len))]

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in _getitem(self, key)
   8554         if len(self._pages) == 1 and 0 < key < self._len:
   8555             index = self._pages[0].index
-> 8556             return self.parent.pages._getitem(index + key)
   8557         return self._pages[key]
   8558 

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in _getitem(self, key, validate, cache, aspage)
   5560 
   5561         pageindex = key if self._index is None else self._index + (key,)
-> 5562         self._seek(key)
   5563         page = tiffpage(self.parent, index=pageindex, keyframe=self._keyframe)
   5564         if validate and validate != page.hash:

/opt/anaconda3/envs/micro-torch/lib/python3.9/site-packages/tifffile/tifffile.py in _seek(self, index, maxpages)
   5418         fh = self.parent.filehandle
   5419         if fh.closed:
-> 5420             raise ValueError('seek of closed file')
   5421 
   5422         if self._indexed or 0 <= index < lenpages:

ValueError: seek of closed file
```
</details>

Im using tifffile version 2021.8.8, dask version 2021.08.1 and zarr version 2.9.3.

Also im not sure whether this is relevant but the failure does seems to depend on the shape of the array. For instance, the same code with `shape =(4,5,6)` or `(3,4,5)` works but `(5,6,7)` breaks in the same way.

Thanks in advance for the help.

**Comment by cgohlke** [comment]
> For instance, the same code with shape =(4,5,6) or (3,4,5) works but (5,6,7) breaks in the same way.

Thanks for reporting! The issue is limited to multi-threaded access of memory-mapable multi-page TIFFs written by tifffile. There are file operations in that code path that are not locked. This will be fixed in the next version.

**Comment by cgohlke** [comment]
Should be fixed in v2021.8.30.

**Comment by thewtex** [comment]
I have encountered this issue (or a similar one) with tifffile 2023.4.12 and newer. Reproducible example here:

https://github.com/thewtex/tiff-to-ome-zarr-with-pixel-type-conversion

**Comment by cgohlke** [comment]
> I have encountered this issue (or a similar one) with tifffile 2023.4.12 and newer. Reproducible example here:

Thanks. It's actually reproducible in a three-liner. Don't know why it never occurred in the many tests or my usage...

**Comment by thewtex** [comment]
@cgohlke thanks for taking a look :pray:

**Comment by cgohlke** [comment]
> I have encountered this issue (or a similar one) with tifffile 2023.4.12 and newer. Reproducible example here:

Fixed in v2024.7.24. It passes the [tiff-to-ome-zarr-with-pixel-type-conversion](https://github.com/thewtex/tiff-to-ome-zarr-with-pixel-type-conversion) example.

**Comment by thewtex** [comment]
@cgohlke outstanding!! Thank you for the rapid fix! :tada: :taco:
